File: 0525-contiguous-array/0525-contiguous-array.java
 class Solution {
    public int findMaxLength(int[] nums) {
        if(nums.length == 1) return 0;
        
        int sum = 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        int ans = Integer.MIN_VALUE;
        map.put(0, -1);
        for(int i = 0; i < nums.length; i++){
            if(nums[i] == 1){
                sum++;
            }else{
                sum--;
            }
            
            if(!map.containsKey(sum)){
                map.put(sum, i);
            }else{
                ans =  Math.max(ans,i - map.get(sum));
            }
            
        }
        if(map.size() == nums.length + 1) return 0;
        return ans;
    }
}

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n = nums.size();
        int ans = 0;
        int ones = 0; // will decrement when zero is found
        
        vector<int> v(n+1);
        
        for(int i = 0; i < n; i++){
            if(nums[i] == 1) ones++;
            if(nums[i] == 0) ones--;
            v[i+1] = ones;
        }
        
        unordered_map<int, int> m;
        for(int i = 0; i <= n; i++){
            if(m.count(v[i])){
                ans = max(ans, i - m[v[i]]);
            }else{
                m[v[i]] = i;
            }
        }
        
        return ans;
    }
};